tests/rptest/scale_tests/many_partitions_test.py

```diff
@@ -245,142 +176,290 @@ def test_many_partitions(self):
+            consumer.stop()
+            consumer.free()
 
-                for node_name, fut in futs.items():
-                    file_count = fut.result()
-                    counts[node_name] = file_count
+    def setUp(self):
+        # defer redpanda startup to the test, it might want to tweak
+        # ResourceSettings based on its parameters.
+        pass
 
-            return counts
+    def _get_fd_counts(self):
+        counts = {}
+        with concurrent.futures.ThreadPoolExecutor(
+                max_workers=len(self.redpanda.nodes)) as executor:
+            futs = {}
+            for node in self.redpanda.nodes:
+                futs[node.name] = executor.submit(
+                    lambda: sum(1 for _ in self.redpanda.lsof_node(node)))
 
-        for node_name, file_count in get_fd_counts().items():
-            self.logger.info(
-                f"Open files before restarts on {node_name}: {file_count}")
+            for node_name, fut in futs.items():
+                file_count = fut.result()
+                counts[node_name] = file_count
+
+        return counts
+
+    def _concurrent_restart(self):
+        """
+        Restart the whole cluster, all nodes in parallel.
+        """
+
+        # Normal restarts are rolling restarts, but because replay takes substantial time,
+        # on an idle system it is helpful to do a concurrent global restart rather than
+        # waiting for each node one by one.
+        with concurrent.futures.ThreadPoolExecutor(
+                max_workers=len(self.redpanda.nodes)) as executor:
+            futs = []
+            for node in self.redpanda.nodes:
+                futs.append(
+                    executor.submit(self.redpanda.restart_nodes,
+                                    nodes=[node],
+                                    start_timeout=self.EXPECT_START_TIME))
+
+            for f in futs:
+                # Raise on error
+                f.result()
+
+    def _single_node_restart(self, scale: ScaleParameters, topic_names: list,
+                             n_partitions: int):
+        """
+        Restart a single node to check stability through the movement of
+        leadership to other nodes, plus the subsequent leader balancer
+        activity to redistribute after it comes back up.
+        """
+
+        node = self.redpanda.nodes[-1]
+        self.logger.info(f"Single node restart on node {node.name}")
+        node_id = self.redpanda.idx(node)
+
+        self.redpanda.stop_node(node)
+
+        # Wait for leaderships to stabilize on the surviving nodes
+        wait_until(
+            lambda: self._node_leadership_evacuated(topic_names, n_partitions,
+                                                    node_id), 30, 1)
+
+        self.redpanda.start_node(node, timeout=self.EXPECT_START_TIME)
+
+        # Heuristic: in testing we see leaderships transfer at about 10
+        # per second.  2x margin for error.  Up to the leader balancer period
```

Comment:
Are you concerned things like "2x margin" will cause tons of flappers? In some earlier tests I did on EC2 I saw leadership transfer at about 3 per second, already outside of the 2x bound.


Is the idea here to catch performance regressions (i.e., if it's worse than 2x we should know about it, because something is wrong)? If not, why not use a much larger threshold to avoid flappers? The primary downside would be a longer time-to-failure, I guess.

On the same topic, I wonder if a bit more sophisticated `wait_until` would be useful, which takes a function which reports some kind of progress metric versus just a boolean, so it can make a decision about whether progress has stalled but also especially log the progress prior to declaring failure, which will be helpful in diagnosing any flappers.

Anyway, I am allergic to flappers which is why I am especially interested in this aspect.

Reply:
> Is the idea here to catch performance regressions (i.e., if it's worse than 2x we should know about it, because something is wrong)? If not, why not use a much larger threshold to avoid flappers? The primary downside would be a longer time-to-failure, I guess.

It's both performance and time to failure. The performance part is to validate the leader balancer's new concurrent update behavior in 22.2 is working properly.

The actual rate here is governed principally by recovery: our leadership transfers are completing when their partition's recovery to the previously-down node completes.

If this becomes a source of failures, we can adjust: anecdotally this wasn't an area I was seeing flakiness up to this point, because the recovery load is pretty well defined by the time it takes to restart the node.  I think the evolution here will be to explicitly wait for recovery to complete first, then wait for leader transfers (which should then be much faster+more deterministic when it's not colliding with recoveries)

Reply:
Thanks for your comments, and good point about validating the concurrent update behavior.

Overall, tests that fail if an expected throughput can't be met would seem to be both (1) quite important and (2) be much less general in where they can run, etc: expecting a specific type of hardware (or at least a minimum spec). I wonder if we should tag them in some way so we can easily run the "rest" of the tests that don't have problem (2) in other places.

It would be nice if we had as many ways as possible to do a scale test w/o this kind of validating of expected throughput too, e.g., measuring some more invariant metric at the storage subsystem, checking concurrency, or running with a "simulated" disk on top of the real disk (i.e., we write to disk as usual but we also feed the IO flow into a model which spits out we expect this series of IO to perform on some disk: though this is kind of hard to make work on very slow IO devices since they necessarily distort the IO timing of future requests).

Reply:
Right, we're inside this implicit bubble of "these tests require dedicated infra _and_ have expectations based on AWS (or even more specifically i3en) ratios of disk/cpu/mem/net).  Making that explicit would be great.